File: packages/shared/src/ipc/contracts/extensionBridge.ts
/**
 * Extension-bridge control capability (Feature 2).
 *
 * Read the local WebSocket bridge's status (bound port, whether an extension is
 * currently paired/connected, and the current pairing token) and regenerate the
 * pairing token. The bridge itself is a loopback WS server in
 * `apps/desktop/src-tauri/src/extension_bridge`; this namespace is only the
 * renderer's control surface over it (show the token in Settings, rotate it).
 */

/** Current bridge status. `port` is `null` when the server failed to bind. */
export interface ExtensionBridgeStatus {
  port: number | null;
  connected: boolean;
  token: string;
}

/** Result of rotating the pairing token (existing sockets must re-pair). */
export interface ExtensionBridgeTokenResult {
  token: string;
}

/**
 * Assisted-autofill opt-in state. When `enabled`, a `profile.get` from the
 * extension returns the user's contact profile so it can fill matching empty form
 * fields on the current page; when off, the desktop refuses. Default OFF.
 */
export interface ExtensionAutofillSetting {
  enabled: boolean;
}

/**
 * AI-answer-assist opt-in state (extension roadmap PR 9) — a SEPARATE opt-in
 * from {@link ExtensionAutofillSetting}: `answer.assist` is billable provider
 * spend (a materially different consent class from the local/free autofill
 * verbs), so it gets its own desktop-enforced gate, default OFF.
 *
 * A bare boolean flag: no provider/model snapshot. A draft resolves the active
 * provider from the backend-owned active-provider store at answer-time (task
 * #16), so a Settings row reads that store (`aiActiveConfig`) for its
 * "Using: X · Y" label rather than a field echoed back here.
 */
export interface ExtensionAiAssistSetting {
  enabled: boolean;
}

/**
 * Auto-track opt-in state (Task #22, auto-track Layer A) — a SEPARATE opt-in
 * from {@link ExtensionAutofillSetting}/{@link ExtensionAiAssistSetting}. When
 * on, the extension arms a gesture submit-watcher that, on a detected form
 * submit, auto-marks the matched `saved` application `applied` (or nudges you
 * to import an untracked one). Default OFF, desktop-enforced: the desktop also
 * re-checks it before honoring an automated write, so a compromised extension
 * can't auto-mark applied without this consent.
 */
export interface ExtensionAutoTrackSetting {
  enabled: boolean;
}

/**
 * Pushed on a 0→1 or →0 transition in the live paired-connection COUNT (not a
 * per-socket event) — the desktop supports multiple browsers sharing one
 * pairing token, each with its own socket, so this only fires when the last
 * one disconnects or the first one (re)connects, never on an intermediate
 * pairing/close while at least one other socket stays open.
 */
export interface ExtensionBridgeChangedEvent {
  connected: boolean;
}

export interface ExtensionBridgeContract {
  status(): Promise<ExtensionBridgeStatus>;
  regenerateToken(): Promise<ExtensionBridgeTokenResult>;
  /** Read the assisted-autofill opt-in (default OFF). */
  autofillEnabled(): Promise<ExtensionAutofillSetting>;
  /** Set + persist the assisted-autofill opt-in; echoes the stored value. */
  setAutofillEnabled(enabled: boolean): Promise<ExtensionAutofillSetting>;
  /** Read the AI-answer-assist opt-in (default OFF). */
  aiAssistEnabled(): Promise<ExtensionAiAssistSetting>;
  /**
   * Set + persist the AI-answer-assist opt-in; echoes the stored value. A bare
   * boolean — the billable-AI consent gate. It no longer snapshots a provider:
   * a draft resolves the active provider from the backend active-provider store
   * at answer-time (task #16), so nothing more needs capturing here.
   */
  setAiAssistEnabled(enabled: boolean): Promise<ExtensionAiAssistSetting>;
  /** Read the auto-track opt-in (default OFF). */
  autoTrackEnabled(): Promise<ExtensionAutoTrackSetting>;
  /** Set + persist the auto-track opt-in; echoes the stored value. */
  setAutoTrackEnabled(enabled: boolean): Promise<ExtensionAutoTrackSetting>;
  /** Subscribe to a live connection-count transition (0→1 / →0). Returns a
   *  sync unsubscribe handle — mirrors `ApplicationsContract.onChanged`. */
  onChanged(handler: (event: ExtensionBridgeChangedEvent) => void): () => void;
}

export const EXTENSION_BRIDGE_CHANNELS = {
  status: 'extensionBridge:status',
  regenerateToken: 'extensionBridge:regenerateToken',
  autofillEnabled: 'extensionBridge:autofillEnabled',
  setAutofillEnabled: 'extensionBridge:setAutofillEnabled',
  aiAssistEnabled: 'extensionBridge:aiAssistEnabled',
  setAiAssistEnabled: 'extensionBridge:setAiAssistEnabled',
  autoTrackEnabled: 'extensionBridge:autoTrackEnabled',
  setAutoTrackEnabled: 'extensionBridge:setAutoTrackEnabled',
} as const;

File: apps/desktop/src-tauri/src/commands/extension_bridge.rs
//! Extension-bridge control commands — the renderer's IPC surface over the local
//! WebSocket bridge ([`crate::extension_bridge`]).
//!
//! `status` (bound port + whether an extension is paired + the current token)
//! and `regenerate_token` (rotate the pairing secret; existing sockets must
//! re-pair), plus two independent opt-in pairs — assisted autofill
//! (`autofill_enabled`/`set_autofill_enabled`) and AI answer-assist
//! (`ai_assist_enabled`/`set_ai_assist_enabled`, a SEPARATE gate: billable
//! provider spend is a different consent class from the local/free autofill
//! verbs). All resolve the managed [`BridgeState`] from app state and return a
//! `serde_json::Value`, matching the neighbouring command style (e.g.
//! `commands::applications`).

use serde_json::{json, Value};
use tauri::{AppHandle, Manager};

use crate::extension_bridge::BridgeState;

/// `{ port, connected, token }`. `port` is `null` when the bridge failed to bind
/// (disabled). When `BridgeState` isn't managed at all (start-up failure before
/// `manage`), report a disabled bridge with an empty token rather than erroring.
#[tauri::command]
pub async fn extension_bridge_status(app: AppHandle) -> Value {
    match app.try_state::<BridgeState>() {
        Some(state) => json!({
            "port": state.port(),
            "connected": state.is_connected(),
            "token": state.token(),
        }),
        None => json!({ "port": Value::Null, "connected": false, "token": "" }),
    }
}

/// Rotate the pairing token and return the new value: `{ token }`. If the bridge
/// state is unavailable, return an empty token (nothing to rotate).
#[tauri::command]
pub async fn extension_bridge_regenerate_token(app: AppHandle) -> Value {
    match app.try_state::<BridgeState>() {
        Some(state) => json!({ "token": state.regenerate_token() }),
        None => json!({ "token": "" }),
    }
}

/// Current assisted-autofill opt-in: `{ enabled }`. Default OFF when the bridge
/// state isn't managed (start-up failure) — the safe state.
#[tauri::command]
pub async fn extension_bridge_autofill_enabled(app: AppHandle) -> Value {
    let enabled = app
        .try_state::<BridgeState>()
        .map(|s| s.autofill_enabled())
        .unwrap_or(false);
    json!({ "enabled": enabled })
}

/// Set (and persist) the assisted-autofill opt-in; echoes the stored value:
/// `{ enabled }`. A no-op returning `false` when the bridge state is unavailable.
#[tauri::command]
pub async fn extension_bridge_set_autofill_enabled(app: AppHandle, enabled: bool) -> Value {
    match app.try_state::<BridgeState>() {
        Some(state) => {
            state.set_autofill_enabled(enabled);
            json!({ "enabled": state.autofill_enabled() })
        }
        None => json!({ "enabled": false }),
    }
}

/// Current AI-answer-assist opt-in: `{ enabled }` — a SEPARATE opt-in from
/// `extension_bridge_autofill_enabled` (billable provider spend is a
/// materially different consent class). No provider/model snapshot is
/// surfaced: a draft resolves the active provider from the backend
/// `AiConfigStore` at answer-time (task #16), so Settings reads that store
/// (`ai_active_config`) for its "Using: X · Y" label. Default OFF when the
/// bridge state isn't managed (start-up failure) — the safe state.
#[tauri::command]
pub async fn extension_bridge_ai_assist_enabled(app: AppHandle) -> Value {
    let enabled = app
        .try_state::<BridgeState>()
        .map(|s| s.ai_assist_enabled())
        .unwrap_or(false);
    json!({ "enabled": enabled })
}

/// Set (and persist) the AI-answer-assist opt-in; echoes the stored value:
/// `{ enabled }`. A bare boolean — the billable-AI consent gate (ADR-0011).
/// It no longer snapshots the renderer's active provider: a draft resolves
/// the active provider from the backend `AiConfigStore` via
/// `Completer::from_active` at answer-time (task #16). A no-op returning
/// `false` when the bridge state is unavailable.
#[tauri::command]
pub async fn extension_bridge_set_ai_assist_enabled(app: AppHandle, enabled: bool) -> Value {
    match app.try_state::<BridgeState>() {
        Some(state) => {
            state.set_ai_assist(enabled);
            json!({ "enabled": state.ai_assist_enabled() })
        }
        None => json!({ "enabled": false }),
    }
}

/// Current auto-track opt-in: `{ enabled }` (Task #22) — default OFF. Gates the
/// extension's gesture submit-watcher (the extension reads it via
/// `autotrack.check`) AND the desktop honoring an AUTO `status.update`. Default
/// OFF when the bridge state isn't managed (start-up failure) — the safe state.
#[tauri::command]
pub async fn extension_bridge_auto_track_enabled(app: AppHandle) -> Value {
    let enabled = app
        .try_state::<BridgeState>()
        .map(|s| s.autotrack_enabled())
        .unwrap_or(false);
    json!({ "enabled": enabled })
}

/// Set (and persist) the auto-track opt-in; echoes the stored value:
/// `{ enabled }`. A bare boolean — the auto-mark-applied consent gate. A no-op
/// returning `false` when the bridge state is unavailable.
#[tauri::command]
pub async fn extension_bridge_set_auto_track_enabled(app: AppHandle, enabled: bool) -> Value {
    match app.try_state::<BridgeState>() {
        Some(state) => {
            state.set_autotrack_enabled(enabled);
            json!({ "enabled": state.autotrack_enabled() })
        }
        None => json!({ "enabled": false }),
    }
}
